File: src/main/java/dev/slne/discord/message/MessageManager.java
package dev.slne.discord.message;

import dev.slne.discord.config.discord.GuildConfig;
import dev.slne.discord.exception.command.CommandException;
import dev.slne.discord.extensions.MemberExtensions;
import dev.slne.discord.spring.feign.dto.WhitelistDTO;
import dev.slne.discord.spring.service.whitelist.WhitelistService;
import dev.slne.discord.ticket.Ticket;
import dev.slne.discord.ticket.member.TicketMember;
import dev.slne.discord.util.TimeUtils;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import javax.annotation.Nonnull;
import lombok.experimental.ExtensionMethod;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.exceptions.ErrorResponseException;
import net.dv8tion.jda.api.interactions.InteractionHook;
import net.dv8tion.jda.api.requests.ErrorResponse;
import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

/**
 * The type Message manager.
 */
@Service
@ExtensionMethod({MemberExtensions.class})
public class MessageManager {

  private static final ComponentLogger LOGGER = ComponentLogger.logger("MessageManager");
  private final EmbedManager embedManager;
  private final JDA jda;
  private final WhitelistService whitelistService;

  /**
   * Private constructor to prevent instantiation.
   */
  private MessageManager(EmbedManager embedManager, JDA jda, WhitelistService whitelistService) {
    this.embedManager = embedManager;
    this.jda = jda;
    this.whitelistService = whitelistService;
  }

  /**
   * Returns an error MessageEmbed with the given title and description.
   *
   * @param title       The title of the embed.
   * @param description The description of the embed.
   * @return The MessageEmbed.
   */
  public static @Nonnull MessageEmbed getErrorEmbed(String title, String description) {
    EmbedBuilder embedBuilder = new EmbedBuilder();

    embedBuilder.setTitle(title);
    embedBuilder.setDescription(description);
    embedBuilder.setColor(EmbedColors.ERROR_COLOR);
    embedBuilder.setTimestamp(TimeUtils.berlinTimeProvider().getCurrentTime());

    return embedBuilder.build();
  }

  @Async
  public CompletableFuture<Void> sendTicketClosedMessages(Ticket ticket) {
    final MessageEmbed closeEmbed = embedManager.buildTicketClosedEmbed(ticket).join();

    if (closeEmbed == null) {
      return CompletableFuture.completedFuture(null);
    }

    final User author = ticket.getTicketAuthorNow();
    final Guild guild = ticket.getGuild();

    if (guild == null) {
      return CompletableFuture.completedFuture(null);
    }

    final GuildConfig guildConfig = GuildConfig.getByGuild(guild);

    if (guildConfig == null) {
      return CompletableFuture.completedFuture(null);
    }

    return CompletableFuture.completedFuture(CompletableFuture.allOf(ticket.getMembers().stream()
        .filter(TicketMember::isActivated)
        .map(ticketMember -> sendTicketClosedMessage(closeEmbed, ticketMember, author, guild,
            guildConfig))
        .toArray(CompletableFuture[]::new)).join());
  }

  @Async
  protected CompletableFuture<Void> sendTicketClosedMessage(
      MessageEmbed closedEmbed,
      @NotNull TicketMember receiver,
      User author,
      Guild guild,
      GuildConfig guildConfig
  ) {
    final User receiverUser = receiver.getMemberNow();

    if (receiverUser == null || receiverUser.equals(jda.getSelfUser())) {
      return CompletableFuture.completedFuture(null);
    }

    final boolean isAuthor = receiverUser.equals(author);
    final Member receiverMember = guild.retrieveMember(receiverUser).complete();
    final boolean isTeamMember = receiverMember.isTeamMember(guildConfig);

    if (!isAuthor && isTeamMember) {
      return CompletableFuture.completedFuture(null);
    }

    try {
      receiverUser.openPrivateChannel()
          .flatMap(channel -> channel.sendMessageEmbeds(closedEmbed))
          .complete();
    } catch (ErrorResponseException e) {
      if (e.getErrorResponse().equals(ErrorResponse.CANNOT_SEND_TO_USER)) {
        return CompletableFuture.completedFuture(null);
      }

      LOGGER.error("Failed to send ticket closed message to user %s".formatted(receiverUser), e);
    }

    return CompletableFuture.completedFuture(null);
  }

  @Async
  public void printUserWlQuery(User user, TextChannel channel) {
    channel.sendTyping().queue();
    final List<WhitelistDTO> whitelists = whitelistService.checkWhitelists(null, user.getId(), null)
        .join();

    try {
      printUserWlQuery(whitelists, user.getName(), channel, null);
    } catch (CommandException e) {
      channel.sendMessage(e.getMessage()).queue();
    }
  }

  public void printUserWlQuery(List<WhitelistDTO> whitelists, String name, TextChannel channel,
      @Nullable InteractionHook hook)
      throws CommandException {
    if (whitelists.isEmpty()) {
      throw new CommandException(
          "Es wurden keine Whitelist Einträge für \"" + name + "\" gefunden.");
    }

    printWlQuery(channel, "\"" + name + "\"", whitelists);

    if (hook != null) {
      hook.deleteOriginal().queue();
    }
  }

  @Async
  public void printWlQuery(TextChannel channel, String title, List<WhitelistDTO> whitelists) {
    title = title.replace("\"", "");
    channel.sendMessage("WlQuery für: `" + title + "`").queue();

    for (final WhitelistDTO whitelist : whitelists) {
      final MessageEmbed embed = getWhitelistQueryEmbed(whitelist).join();
      channel.sendMessageEmbeds(embed).queue();
    }
  }

  @Async
  public CompletableFuture<MessageEmbed> getWhitelistQueryEmbed(@NotNull WhitelistDTO whitelist) {
    final EmbedBuilder builder = new EmbedBuilder()
        .setTitle("Whitelist Query")
        .setFooter("Whitelist Query", jda.getSelfUser().getAvatarUrl())
        .setDescription("Whitelist Informationen")
        .setColor(EmbedColors.WL_QUERY)
        .setTimestamp(ZonedDateTime.now(TimeFormatter.EUROPE_BERLIN));

    final String minecraftName = whitelist.getMinecraftName();
    final String twitchLink = whitelist.getTwitchLink();
    final UUID uuid = whitelist.getUuid();
    final User discordUser = whitelist.getDiscordUserNow();
    final User addedBy = whitelist.getAddedByNow();

    if (minecraftName != null) {
      builder.addField("Minecraft Name", minecraftName, true);
    }

    if (twitchLink != null) {
      builder.addField("Twitch Link", twitchLink, true);
    }

    if (discordUser != null) {
      builder.addField("Discord User", discordUser.getAsMention(), true);
    }

    if (addedBy != null) {
      builder.addField("Added By", addedBy.getAsMention(), true);
    }

    builder.addField("UUID", uuid.toString(), false);

    return CompletableFuture.completedFuture(builder.build());
  }
}

File: src/main/kotlin/dev/slne/discord/message/MessageManager.kt
package dev.slne.discord.message

import dev.minn.jda.ktx.coroutines.await
import dev.minn.jda.ktx.messages.Embed
import dev.minn.jda.ktx.messages.MessageCreate
import dev.slne.discord.discord.interaction.command.getGuildConfig
import dev.slne.discord.exception.command.CommandException
import dev.slne.discord.exception.command.CommandExceptions
import dev.slne.discord.message.EmbedColors.ERROR_COLOR
import dev.slne.discord.message.EmbedColors.WL_QUERY
import dev.slne.discord.persistence.external.Whitelist
import dev.slne.discord.persistence.service.user.UserService
import dev.slne.discord.persistence.service.whitelist.WhitelistService
import dev.slne.discord.ticket.Ticket
import dev.slne.discord.util.memberOrNull
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel
import net.dv8tion.jda.api.interactions.InteractionHook
import org.springframework.stereotype.Component
import java.time.ZonedDateTime

@Component
class MessageManager(
    private val whitelistService: WhitelistService,
    private val userService: UserService,
    private val jda: JDA
) {

    suspend fun sendTicketClosedMessages(ticket: Ticket): Message? =
        ticket.thread?.sendMessage(MessageCreate {
            embeds += EmbedManager.buildTicketClosedEmbed(ticket)
        })?.await()

    suspend fun sendTicketClosedUserPrivateMessage(ticket: Ticket) {
        ticket.author.await()?.openPrivateChannel()?.await()?.sendMessage(MessageCreate {
            embeds += EmbedManager.buildTicketClosedUserPrivateMessageEmbed(
                ticket,
                owner = true
            )
        })?.await()

        val guild = ticket.guild ?: return
        val thread = ticket.thread ?: return

        for (ticketMember in thread.threadMembers) {
            if (hasMemberReceiveClosePmNegatePermission(
                    ticketMember.user,
                    guild
                ) || ticketMember.user.isBot
            ) {
                continue
            }

            ticketMember.user.openPrivateChannel().await()?.sendMessage(MessageCreate {
                embeds += EmbedManager.buildTicketClosedUserPrivateMessageEmbed(ticket)
            })?.await()
        }
    }

    private fun hasMemberReceiveClosePmNegatePermission(user: User, guild: Guild): Boolean {
        val roles = user.memberOrNull(guild)?.roles ?: return false
        val guildTeamRoles =
            guild.getGuildConfig()?.discordGuild?.roles?.flatMap { it.discordRoleIds }
                ?: return false

        return roles.any { it.id in guildTeamRoles }
    }


    suspend fun printUserWlQuery(user: User, channel: MessageChannel) {
        channel.sendTyping().await()
        val whitelists = whitelistService.findWhitelists(null, user.id, null)

        try {
            printUserWlQuery(whitelists, user.name, channel, null)
        } catch (exception: CommandException) {
            channel.sendMessage("${exception.message}").await()
        }
    }

    suspend fun printUserWlQuery(
        whitelists: List<Whitelist>, name: String, channel: MessageChannel, hook: InteractionHook?
    ) {
        if (whitelists.isEmpty()) {
            throw CommandExceptions.WHITELIST_QUERY_NO_ENTRIES.create(name)
        }

        printWlQuery(channel, "\"" + name + "\"", whitelists)

        hook?.deleteOriginal()?.await()
    }

    private suspend fun printWlQuery(
        channel: MessageChannel,
        title: String,
        whitelists: List<Whitelist>
    ) {
        channel.sendMessage(translatable("whitelist.query.start", title.replace("\"", "")))
            .await()

        for (whitelist in whitelists) {
            channel.sendMessageEmbeds(getWhitelistQueryEmbed(whitelist)).await()
        }
    }

    fun getErrorEmbed(title: String?, description: String?) = Embed {
        this.title = title
        this.description = description
        color = ERROR_COLOR
        timestamp = ZonedDateTime.now()
    }

    suspend fun getWhitelistQueryEmbed(whitelist: Whitelist) = Embed {
        title = translatable("whitelist.query.embed.title")
        footer {
            name = translatable("whitelist.query.embed.footer")
            iconUrl = jda.selfUser.avatarUrl
        }
        description = translatable("whitelist.query.embed.description")
        color = WL_QUERY
        timestamp = ZonedDateTime.now()

        val minecraftName = userService.getUsernameByUuid(whitelist.uuid)
        val twitchLink = whitelist.twitchLink
        val uuid = whitelist.uuid
        val discordUser = whitelist.user?.await()
        val addedBy = whitelist.addedBy?.await()

        field {
            name = translatable("whitelist.query.embed.field.uuid")
            value = uuid.toString()
            inline = false
        }

        if (minecraftName != null) {
            field {
                name = translatable("whitelist.query.embed.field.minecraft-name")
                value = "`${minecraftName}`"
            }
        }

        field {
            name = translatable("whitelist.query.embed.field.twitch-name")
            value = "[${twitchLink}](${whitelist.clickableTwitchLink})"
        }

        if (discordUser != null) {
            field {
                name = translatable("whitelist.query.embed.field.discord-user")
                value = discordUser.asMention
            }
        }

        if (addedBy != null) {
            field {
                name = translatable("whitelist.query.embed.field.added-by")
                value = addedBy.asMention
            }
        }

        field {
            name = translatable("whitelist.query.embed.field.blocked")
            value = if (whitelist.blocked) translatable("common.yes") else translatable("common.no")
        }
    }

    fun buildMemberAddedMessage(
        member: Member,
        executor: User
    ) = MessageCreate {
        content = member.asMention
        embed {
            title = translatable("interaction.command.ticket.member.embed.title")
            description = translatable("interaction.command.ticket.member.embed.description")
            timestamp = ZonedDateTime.now()
            color = EmbedColors.ADD_TICKET_MEMBER
            footer {
                name = translatable(
                    "interaction.command.ticket.member.embed.footer",
                    executor.name
                )
                iconUrl = executor.getAvatarUrl()
            }
        }
    }
}
